transport/httptransport/http_transport.go

```diff
@@ -0,0 +1,241 @@
+		if err := t.pub.Emit(types.TransportEvent{
+			Status:          transferstatus.Errored,
+			IsTerminalState: true,
+			Error:           err,
+		}); err != nil {
+			log.Errorw("failed to publish transport error", "dealUuid", t.dealInfo.DealUuid, "err", err)
+		}
+	}
+
+	// construct request
+	req, err := http.NewRequest("GET", t.tInfo.URL, nil)
+	if err != nil {
+		failTransfer(fmt.Errorf("failed to create http req: %w", err))
+		return
+	}
+
+	// add range req to start reading from the last byte we have in the output file
+	fi, err := os.Stat(t.dealInfo.OutputFile)
+	if err != nil {
+		failTransfer(fmt.Errorf("failed to stat output file: %w", err))
+		return
+	}
+
+	req.Header.Set("Range", fmt.Sprintf("bytes=%d-", fi.Size()))
+	// init the request with the transfer context
+	req = req.WithContext(ctx)
+	// open output file in append-only mode for writing
+	of, err := os.OpenFile(t.dealInfo.OutputFile, os.O_APPEND|os.O_WRONLY, 0644)
+	if err != nil {
+		failTransfer(fmt.Errorf("failed to open output file: %w", err))
+		return
+	}
+
+	// start the http transfer
+	remaining := t.dealInfo.DealSize - fi.Size()
+	if err := t.doHttp(req, of, remaining); err != nil {
+		// do not resume transfer if context has been cancelled
+		if ctx.Err() == context.Canceled || ctx.Err() == context.DeadlineExceeded {
+			if err := t.pub.Emit(types.TransportEvent{
+				Status:          transferstatus.Cancelled,
+				IsTerminalState: true,
+				Error:           fmt.Errorf("transfer context err: %w", ctx.Err()),
+			}); err != nil {
+				log.Errorw("failed to publish transport error", "dealUuid", t.dealInfo.DealUuid, "err", err)
+			}
+			return
+		}
+
+		// TODO: resumption
+		return
+	}
+
+	// --- http request finished successfully. see if we got the number of bytes we expected.
+
+	// if the number of bytes we've received is not the same as the deal size, we have a failure.
+	if t.nBytesReceived != t.dealInfo.DealSize {
+		failTransfer(fmt.Errorf("mismatch in dealSize vs received bytes, dealSize=%d, received=%d", t.dealInfo.DealSize, t.nBytesReceived))
+		return
+	}
+
+	// otherwise we're good ! notify the subscriber.
+	if err := t.pub.Emit(types.TransportEvent{
+		Status:          transferstatus.Finished,
+		NBytesReceived:  t.nBytesReceived,
+		IsTerminalState: true,
+	}); err != nil {
+		log.Errorw("failed to publish transport event", "dealUuid", t.dealInfo.DealUuid, "err", err)
+	}
+}
+
+func (t *transfer) doHttp(req *http.Request, dst io.Writer, toRead int64) error {
+	// send http request and validate response
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to send  http req: %w", err)
+	}
+	// we should either get back a 200 or a 206 -> anything else means something has gone wrong and we return an error.
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusPartialContent {
+		resp.Body.Close() // nolint
+		return fmt.Errorf("http req failed: %d", resp.StatusCode)
+	}
+	defer resp.Body.Close()
```

Comment:
nit: move this `defer resp.Body.Close()` to line 186 and remove `defer resp.Body.Close()` on line 188